File: MoonApp/packages/react-native-nitro-moon/ios/Session/SessionBody.swift
import Foundation
import SwiftCBOR
import SwiftyJSON

struct SessionBody: Codable {
    // 设备唯一标识
    let deviceUID   : String
    // AES-GCM 验证标签
    let tag: Data
    // 共享密钥盐，AES-GCM随机值
    let nonce: Data
    // 密文数据
    let body: Data
    // 附加数据
    var payload: Data?
    
    /// 获取负载数据
    func getPayload() -> [String: Any]? {
        guard let payload = payload else { return nil }
        return CBOR.decodeData(payload) as? [String: Any]
    }
    
    /// 设置负载数据
    mutating func setPayload(_ payload: [String: Any]) {
        self.payload = CBOR.encodeData(payload)
    }
    
    /// 转换成字典数据
    func toDict() -> [String: Any] {
        var dict: [String: Any] = [:]
        dict["deviceUID"]   = deviceUID
        dict["tag"]         = tag
        dict["nonce"]       = nonce
        dict["body"]        = body
        
        if let payloadData = payload {
            dict["payload"] = payloadData
        }
        
        return dict
    }
    
    /// 转换成 JSON
    func toJSON() -> JSON {
        let dict = toDict()
        return JSON(dict)
    }
    
    /// 转换成 CBOR
    func toCBORData() -> Data {
        let dict = toDict()
        return CBOR.encodeData(dict)
    }
    
    /// 从 Dict 创建
    static func fromDict(_ dict: [String: Any]) -> SessionBody? {
        guard let deviceUID = dict["deviceUID"] as? String,
              let tagBase64 = dict["tag"] as? String,
              let tag = Data(base64Encoded: tagBase64),
              let nonceBase64 = dict["nonce"] as? String,
              let nonce = Data(base64Encoded: nonceBase64),
              let bodyBase64 = dict["body"] as? String,
              let body = Data(base64Encoded: bodyBase64)
        else {
            return nil
        }
        
        var sessionBody = SessionBody(
            deviceUID: deviceUID, tag: tag, nonce: nonce, body: body
        )
            
        if let payloadBase64 = dict["payload"] as? String,
           let payload = Data(base64Encoded: payloadBase64) {
            sessionBody.payload = payload
        }
        
        return sessionBody
        
    }
    
    /// 从 JSON 创建
    static func fromJSON(_ data: JSON) -> SessionBody? {
        guard let deviceUID = data["deviceUID"].string,
              let tagBase64 = data["tag"].string,
              let tag = Data(base64Encoded: tagBase64),
              let nonceBase64 = data["nonce"].string,
              let nonce = Data(base64Encoded: nonceBase64),
              let bodyBase64 = data["body"].string,
              let body = Data(base64Encoded: bodyBase64)
        else {
            return nil
        }
        
        var sessionBody = SessionBody(
            deviceUID: deviceUID, tag: tag, nonce: nonce, body: body
        )
        
        if let payloadBase64 = data["payload"].string, let payload = Data(base64Encoded: payloadBase64) {
            sessionBody.payload = payload
        }
        
        return sessionBody
    }
    
    /// 从 CBOR 创建
    static func fromCBOR(_ data: Data) -> SessionBody? {
        guard let body = CBOR.decodeData(data), let dict = body as? [String: Any] else {
            return nil
        }
        
        guard let deviceUID = dict["deviceUID"] as? String,
              let tag       = dict["tag"] as? Data,
              let nonce     = dict["nonce"] as? Data,
              let body      = dict["body"] as? Data
        else {
            return nil
        }
        
        var sessionBody = SessionBody(
            deviceUID: deviceUID, tag: tag, nonce: nonce, body: body
        )
            
        if let payload = dict["payload"] as? Data {
            sessionBody.payload = payload
        }
        
        return sessionBody
        
    }
    
}

File: MoonWebsocket/src/Session/SessionBody.ts
import { decode, encode } from "cbor";
import { createError, parseBuffer  } from "../utils";
/**
 * 会话加密内容
 */
export interface SessionBody {
    // 设备或客户端唯一标识
    deviceUID: string;
    // 随机盐
    nonce: Buffer;
    // AES-GCM 验证标签
    tag: Buffer;
    // 加密内容
    body: Buffer;
    // 负载数据
    payload?: Buffer
}


interface SessionBodyMixin extends SessionBody {
    getPayload(): Record<string, any> | null;
    // 设置负载数据
    setPayload(payload: Record<string, any>): void;
    // 转换成 JSON
    toJSON(): Record<string, any>
}


export namespace SessionBody {
    
    /**
     * 创建会话数据
     * @param data 
     */
    export function create(data: SessionBody): SessionBodyMixin {
        return {
            ...data,
            // 获取负载数据
            getPayload() {
                const { payload } = this
                if (payload && Buffer.isBuffer(payload)) {
                    return decode(payload)
                }
                return null
            },

            // 设置负载数据
            setPayload(payload: Record<string, any>) {
                if ( payload === null || typeof payload !== 'object' || Array.isArray(payload) ) {
                    throw new TypeError('Payload must be a plain object')
                }
                if (Object.getPrototypeOf(payload) !== Object.prototype) {
                    throw new Error('Payload must be a plain object')
                }
                this.payload = encode(payload)
            },

            // 转换成 JSON
            toJSON(): Record<string, any> {
                const { deviceUID, nonce, tag, body, payload } = this

                if (typeof deviceUID !== 'string') {
                    throw new Error('Invalid deviceUID.')
                }

                for (const [key, val] of Object.entries({ nonce, tag, body })) {
                    if (!Buffer.isBuffer(val)) {
                        throw new Error(`Invalid ${key}, must be buffer.`)
                    }
                }

                const json: Record<string, any> = {
                    deviceUID,
                    nonce   : nonce.toString('base64'),
                    tag     : tag.toString('base64'),
                    body    : body.toString('base64'),
                }

                if (payload != undefined) {
                    if (!Buffer.isBuffer(payload)) {
                        throw new Error('Invalid payload, must be Buffer')
                    }
                    json.payload = payload.toString('base64')
                }
                return json
            }
        }
    }

    /**
     * 从 JSON 创建会话数据
     * @param data 
     */
    export function fromJSON(data: unknown) {
        if (!data || typeof data !== 'object' || Array.isArray(data)) {
            throw new createError(412, '数据格式不正确')
        }
        let { deviceUID, nonce, tag, body, payload } = data as Record<string, any>

        if (typeof deviceUID !== 'string') {
            throw new createError(412, 'deviceUID 格式不正确')
        }
   
        const bodyData: SessionBody = {
            deviceUID   : deviceUID,
            nonce       : parseBuffer(nonce),
            tag         : parseBuffer(tag),
            body        : parseBuffer(body)
        }

        if (payload != undefined) {
            bodyData.payload = parseBuffer(payload)
        }

        return SessionBody.create(bodyData)
    }
}